## Combinatorial Optimization

#### Assignment IV

***Heuristics for the Knapsack Problem - Solving Multiple Instances***

**Instance:** C500L700

Guilherme Cadori

09/07/2023


In [1]:
# Importing required libraries
import numpy as np


#### 1. Read the instance

In [2]:
# Reading the target file
# GC: the code template below was provided in class and was adapted by GC (2023-07-09)

filename = "C:\gccadori\Lab4_CO\Knap_C500I700.dat"  # Substitua pelo nome do arquivo real

# Variáveis para armazenar os dados lidos
linha = None
coluna = None
cost = None
rhs = None
matrix = None

# Função auxiliar para converter uma linha de valores em uma lista de floats
def parse_values(line):
    return [float(value) for value in line.strip().split()]

# Ler o arquivo e extrair os dados
with open(filename, "r") as file:
    lines = file.readlines()

# Ler o escalar
linha = int(lines[0].strip())
coluna = int(lines[2].strip())

# Ler o vetor
cost = np.array(parse_values(lines[5]))

# Ler a matriz
end = 7 + linha
matrix_lines = [line.strip() for line in lines[7:end]]
matrix = np.array([parse_values(line) for line in matrix_lines],dtype=object)

# Ler vetor de RHS
rhs = np.array(parse_values(lines[end+1]))

# Imprimir os dados lidos
print("\n linhas:", linha)
print("\n colunas:", coluna)
print("\n cost:", cost)
print("\n rhs:", rhs)
print("\n Matrix:")
for row in matrix:
    print(row)



 linhas: 500

 colunas: 700

 cost: [3.0930e-01 4.8131e-01 2.9689e-01 8.9500e-01 2.8789e-01 5.9448e-01
 4.3145e-01 4.3182e-01 6.5954e-01 8.8139e-01 4.8244e-01 4.4517e-01
 9.8283e-01 4.7097e-01 5.1865e-01 9.3413e-01 8.5962e-01 7.1118e-01
 7.3681e-01 5.2363e-01 7.3332e-01 9.7169e-01 2.6586e-01 2.4920e-01
 3.0434e-01 2.0540e-02 2.7340e-01 9.8127e-01 1.8315e-01 2.6742e-01
 5.9633e-01 4.8046e-01 9.1080e-02 7.6191e-01 4.4426e-01 6.4972e-01
 8.9795e-01 7.9562e-01 9.4167e-01 5.9878e-01 6.3169e-01 8.0650e-01
 9.2494e-01 3.8686e-01 3.7680e-02 3.3572e-01 4.6970e-01 2.3685e-01
 7.2448e-01 3.7648e-01 5.5049e-01 1.5585e-01 2.8893e-01 9.8999e-01
 8.4240e-01 2.6936e-01 1.3476e-01 9.6464e-01 6.8671e-01 5.7274e-01
 1.1720e-02 6.2900e-03 7.2460e-01 3.3172e-01 2.4149e-01 7.6667e-01
 4.7543e-01 6.3062e-01 7.5905e-01 2.8470e-01 8.7038e-01 5.4782e-01
 2.0883e-01 7.4625e-01 2.4571e-01 5.8581e-01 7.7126e-01 4.9162e-01
 6.5169e-01 3.8500e-03 6.9020e-01 2.1240e-01 7.5472e-01 5.5324e-01
 2.3680e-01 8.1557e-01 3.

[22.68588 6.59206 17.82618 4.66096 11.76819 12.91333 9.4042 6.38128
 0.09413 7.09957 22.43092 21.43106 16.83706 5.44697 22.14632 13.28718
 17.65684 8.52782 2.07783 22.02682 4.68649 15.84595 22.82161 12.84155
 2.87018 14.09754 12.38031 0.87315 24.9591 12.5509 17.95708 4.73143
 21.10467 6.14545 11.52295 16.17103 11.48884 17.98751 16.01815 17.09338
 13.44766 14.86259 20.51294 10.95869 7.78291 7.41865 10.18377 8.60603
 16.61985 4.84426 17.5394 9.68423 12.87252 23.44661 17.15234 4.30984
 10.45313 10.69263 11.10094 23.53582 16.57923 22.04637 8.40398 20.65864
 9.75059 3.16125 6.11972 4.09628 21.2432 9.53413 15.10715 5.91386 19.17681
 4.61416 0.1222 3.86704 18.3584 24.65885 16.33555 1.62356 12.20481 1.29301
 6.66951 19.41694 15.45815 5.04379 21.04706 12.91146 2.85082 13.76926
 20.00677 3.82894 3.07902 24.15352 23.23398 18.53959 19.80932 10.30594
 11.91858 15.52596 19.86604 13.5131 14.74902 11.81898 16.62856 1.25665
 20.57642 2.91959 19.61424 6.49804 12.55251 19.99602 23.05879 24.15205
 23.5537

[22.2893 16.2909 1.20565 13.29139 13.39352 4.94151 1.88116 16.58167
 13.17003 23.71178 23.82411 11.86878 3.61023 2.16428 0.03218 15.89799
 22.4986 8.92887 17.58003 17.50384 12.0253 9.27697 18.02838 3.02428
 4.09749 16.455 9.16045 9.74185 6.3458 3.88661 22.18683 19.01969 13.96997
 18.30889 17.42994 20.03822 7.38984 0.98528 9.63275 22.67855 8.40873
 0.52126 10.80214 1.55052 9.6609 20.81701 21.46825 16.91137 4.33749
 0.22793 5.89486 24.86349 5.6077 23.54837 2.54586 13.22134 11.0184 11.2835
 16.80637 14.61801 9.91382 21.5461 0.24689 24.4005 24.17021 3.71227
 17.17594 1.10402 5.27376 11.06606 12.26192 11.00409 20.69034 17.62283
 11.94655 10.68081 12.43228 24.26426 9.38783 6.30631 15.14671 20.71931
 4.51418 19.78429 14.52335 18.91195 3.16961 21.65196 4.50463 9.30987
 20.96548 16.81394 16.80336 14.06671 19.18642 16.20951 8.17978 2.53733
 19.8898 12.83168 12.08322 7.68908 5.38959 7.79734 24.97309 22.73701
 15.97124 3.57592 0.42778 14.69412 14.07772 4.23447 18.69148 22.64869
 6.5203 11.68544 22

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



#### 2.	Construct an initial feasible solution

In [3]:
# Renaming variables for testing purposes
c = cost

A = matrix

b = rhs


In [4]:
# Providing an initial feasible solution

def randomStart(nVars, A, b, c):
    # Creating variable names x1, ..., x_nVars
    variable_names = ['x{}'.format(i) for i in range(1, nVars+1)]

    while True:
        np.random.seed(12345)
        x = np.random.randint(2, size=nVars)
        
        if np.all(A @ x <= b):
            FO = x @ c
            print('The initial feasible solution is:')   
            print('Objective function value:', round(FO, 2))
            return


In [5]:
# Testing the function
randomStart(nVars = coluna, A = A, b = b, c = c)


The initial feasible solution is:
Objective function value: 176.2


#### 3.	Program the heuristic you proposed in the previous assingment

In [6]:
# Implementing the heuristic algorithm proposed in Lab III

def naiveKnapsack(nVars, items_c, capacity_b, A):
    # Sort the items in descending order of value
    sorted_items = sorted(range(nVars), key=lambda i: items_c[i], reverse=True)

    # Initialize variables
    selected_items = np.zeros(nVars, dtype=int)
    total_weight = np.zeros_like(capacity_b, dtype=items_c.dtype)
    total_value = 0

    # Greedy selection process
    for idx in sorted_items:
        weight = A[:, idx]
        if np.all(total_weight + weight <= capacity_b):
            # Add the item to the knapsack
            selected_items[idx] = 1
            total_weight += weight.astype(total_weight.dtype)  # Ensure matching data type
            total_value += items_c[idx]

    # Print the selected items and total value
    print('The initial feasible solution is:')
    print('Objective function value:', round(total_value, 2))

    # Return the complete list of items
    return


In [7]:
# Testing the function
naiveKnapsack(nVars = coluna, items_c = c, capacity_b = b , A = A)


The initial feasible solution is:
Objective function value: 306.22


#### 4.	Program a Local Search 

In [8]:
# Local Search metaheuristic implementation
def LocalSearch(nVars, A, b, c, nStarts=1, max_iterations=1000):
    import numpy as np

    # Creating variable names
    variable_names = ['x{}'.format(i) for i in range(1, nVars + 1)]

    best_solution = None
    best_value = -np.inf

    # Escaping Local Optima - Multiple Starting Solutions
    for n in range(1, nStarts + 1):
        # Producing an initial feasible solution
        while True:
            x = np.random.randint(2, size=nVars)
            if np.all(A @ x <= b):
                FO = x @ c
                break  # Exit the loop once an initial feasible solution is found

        # Setting initial search parameters
        s = FO
        s_new = 0
        n_iterations = 0
        best_iteration = 0

        # Termination criteria
        while n_iterations - best_iteration < max_iterations:
            n_iterations += 1

            # Neighbor Selection: Swap Two - First Improvement Strategy
            def generate_neighbor_solution(x):
                x_new = x.copy()
                idx1, idx2 = np.random.choice(len(x), size=2, replace=False)
                x_new[idx1], x_new[idx2] = x_new[idx2], x_new[idx1]
                return x_new

            # Generating neighbor solution
            x_new = generate_neighbor_solution(x)

            # Evaluate solution
            if np.all(A @ x_new <= b):
                FO_new = x_new @ c
            else:
                FO_new = -np.inf

            # Update the current solution
            if FO_new > s_new:
                x = x_new
                s_new = FO_new
                best_iteration = n_iterations

                # Check if the current solution is the best found so far
                if s_new > best_value:
                    best_solution = x.copy()
                    best_value = s_new

    # Print the best solution found
    print('The best solution found:')
    print('Objective function value:', round(best_value, 2))

    return


#### 5.	Running the heuristic for the given instance

In [10]:
import time

# Record starting time
start_time = time.time()

# Running the proposed heuristic for this instance
LocalSearch(nVars = coluna, A = A, b = b, c = c, 
            nStarts = 12,
            max_iterations = 30)

# Record ending time
end_time = time.time()

# Calculate the elapsed time
elapsed_time = round(end_time - start_time, 2)

# Print the elapsed time
print("Elapsed time:", elapsed_time, "seconds")


The best solution found:
Objective function value: 224.21
Elapsed time: 171.79 seconds


**Test Run Results**

The best solution found:

- Objective function value: 224.21
- Elapsed time: 171.79 seconds


### End